File: leetcode/760-find-anagram-mappings/main.go
package main

/*
	naive approach
	- for each word in A, search for the counter part index in B
	Time		O(n^2)
	Space		O(1)
*/

/*
	1st approach
	- hashtable
	Time		O(n)
	Space		O(n)
	0ms beats 100%
*/
func anagramMappings(A []int, B []int) []int {
	ht := make(map[int]int)
	for i := 0; i < len(B); i++ {
		ht[B[i]] = i
	}
	res := []int{}
	for i := 0; i < len(A); i++ {
		res = append(res, ht[A[i]])
	}
	return res
}

func main() {

}

File: leetcode/760-find-anagram-mappings/main.py
from collections import defaultdict

"""
    1st: hashtable

	Time		O(N)
	Space		O(N)
	36 ms, faster than 49.57%
"""
class Solution:
    def anagramMappings(self, A: List[int], B: List[int]) -> List[int]:
        ht = defaultdict(list)
        for i in range(len(B)):
            num = B[i]
            ht[num].append(i)
        res = []
        for num in A:
            top = ht[num].pop()
            res.append(top)
        return res